### PR TITLE
serialization functions between Random.State.t and Bytes.t

### LINKED ISSUE #11322: Add functionality for serialization and deserialization of Random.State
Certain applications (simulations, games, etc) could benefit from the ability to save the current state of the RNG and reload and continue from where they were. It's not enough to have the same seed, because that would require keeping track of how many random numbers have been generated so far, and then looping over the RNG and generating that same number of random numbers, which is inefficient and inelegant. We have functions to load and save the Random.State, but the actual state is opaque, making it impossible to serialize or deserialize unless we use Marshal.

### PR DESCRIPTION
**Pull request by gasche**
closes #11322

### CODE REVIEW

**Comment by gasche** [comment]
```ocaml
  val to_bytes : t -> Bytes.t
  (** Serializes the PRNG state to an array of bytes.

      Note that the serialization format may differ across OCaml
      versions.

      @since 5.1.0
  *)

  val of_bytes : Bytes.t -> t
  (** Deserializes an array of bytes representing a PRNG state.

      Note that the serialization format may differ across OCaml
      versions.

      @raise Invalid_argument if the input is in the expected format.

      @since 5.1.0
  *)
``

File: stdlib/random.mli [hunk 1]
```diff
@@ -148,6 +148,25 @@ module State : sig
...
+      @since 5.1.0
+  *)
+
+  val of_bytes : Bytes.t -> t
+  (** Deserialize an array of bytes representing a PRNG state.
+
+      Note that the serialization format may differ across OCaml
+      versions.
+
+      @raise Invalid_argumnt if the input is in the expected format.
```
Comment (dbuenzli) [inline stdlib/random.mli]: `Invalid_argmu_e_nt` … is *_not_* in the expected format.

Note that since there is no function to check if the given `Bytes.t` are in the expected format this will lead to people having to catch `Invalid_argument` which they are not supposed to. I'd say either raise `Failure` or provide a validation function.

**Review by dbuenzli — COMMENTED** [review]
Personally I would rather take and return a `string`. That way you can avoid having to write the ownership details I'm going to ask you to write in the doc strings.

File: stdlib/random.ml [hunk 2]
```diff
@@ -40,6 +40,35 @@ module State = struct
...
+    for i = 0 to 3 do
+      Bytes.set_int64_le buf (preflen + i * 8) (Array1.get s i)
+    done;
+    buf
+
+  let of_bytes buf =
+    let prefix = serialization_prefix in
+    let preflen = serialization_prefix_len in
+    if Bytes.length buf <> preflen + (4 * 8)
+       || Bytes.sub_string buf 0 preflen <> prefix
```
Comment (hhugo) [inline stdlib/random.ml]: `Bytes.start_with`

File: stdlib/random.mli (hunk 1 above)
Comment (gasche) [reply stdlib/random.mli]: Good point, let's move to `Failure`.

File: stdlib/random.ml (hunk 2 above)
Comment (gasche) [reply stdlib/random.ml]: That's a good idea, I tried, it does not work:
1. the size check is more precise than what `starts_with` does (I check for both the prefix and the payload size), so I would need a size check anyway
2. `Bytes.starts_with` takes a `Bytes.t` as prefix parameter, I have a string, and adding a conversion obscures things. (I guess we should have `{starts,ends}_with_string`.

In the end it was simpler and clearer to leave the code as-is. Thanks anyways :-)

**Comment by gasche** [comment]
@dbuenzli it's the usual question of whether "immutable bytes" should use `bytes` or `string`. I'm not convinced (yet?) that string is better, in particular because the resulting payload is not printable, it may contain arbitrary bytes. (I would have chosen `string` if it was an hexadecimal or base64 encoding for example.)

I'm not sure what ownership details are worth specifying. The result of either functions is "separate" (no aliasing) from their input, but I don't see why one would assume otherwise. I reworded the documentation to say: "into a fresh byte sequence", "into a fresh PRNG state" -- I was inspired by the documentation of `copy`. Does that work for you?

**Comment by xavierleroy** [comment]
> It's the usual question of whether "immutable bytes" should use bytes or string

String, in my opinion.  The bytes / string distinction is about mutability.  Nobody said that strings must be human-readable.  Digests (from the stdlib Digest module) are strings.

Playing devil's advocate: why add marshaling functions for type  `Random.State.t` and not for other abstract types in the stdlib, e.g. sets or maps?  What's wrong with the generic marshaler, exactly?

File: stdlib/random.mli [hunk 3]
```diff
@@ -148,6 +148,26 @@ module State : sig
       correlation.  Both PRNGs can be split later, arbitrarily many times.
       @since 5.0.0 *)
 
+  val to_bytes : t -> Bytes.t
+  (** Serializes the PRNG state into a fresh byte sequence.
+
+      Note that the serialization format may differ across OCaml
+      versions.
```
Comment (xavierleroy) [inline stdlib/random.mli:158]: Then it's not any better than the generic marshaling facility.  If you don't give more guarantees than this, what's the purpose?

**Comment by dbuenzli** [comment]
> I'm not convinced (yet?)

Ok… 

> it's the usual question of whether "immutable bytes" should use bytes or string.

Personally I think this question has been resolved for a long time by now. Immutable bytes should simply always be `string`. The idea that `string` equates with text in OCaml is wrong. `string` is simply a sequence of immutable bytes.

You can make it very clear from the API what the functions think about the ownership of these bytes there's no reason not to have it in types. Even more so that the client is unlikely to be interested in mutating them.

Just to pick on one binary-to-text serialization format, `base64`, every API I can think of ([1][1], [2][2], [3][3], [4][4]) will use `string` as an interface. And to pick one binary serialization format, `cbor`, two APIs ([5][5], [6][6]) I'm aware of, are unsurprsingly using… etc.

Your API just makes it more cumbersome to use with the tools of the eco-system. 

Does that convince you ? 

[1]: http://mirage.github.io/ocaml-base64/base64/Base64/index.html
[2]: https://erratique.ch/software/webs/doc/Webs/Http/Base64/index.html
[3]: https://ocaml.org/p/dream/1.0.0~alpha4/doc/Dream/index.html#val-to_base64url
[4]: https://ocaml.org/p/orsetto/1.0.3/doc/Cf_base64/Std/index.html
[5]: https://github.com/ygrek/ocaml-cbor/blob/480a743fd5a03237bc488a7aec72e102894edee9/src/CBOR.mli#L15
[6]: https://ocaml.org/p/orsetto/1.0.3/doc/Cbor_encode/index.html#val-octets

File: stdlib/random.mli (hunk 3 above)
Comment (gasche) [reply stdlib/random.mli:158]: The use-case described in #11322 is a multiplayer game where, say, the server would send its current PRNG state in its network packets to clients to make sure that the local and shared views of the world remain consistent. (Or for clients just joining the game, etc.) In this use-case you cannot use `Marshal`, because a malicious or corrupted or stale input will segfault the program or lead to a remote code execution. The extra feature provided by these simple functions is dynamic input validation.

**Comment by gasche** [comment]
@dbuenzli I'm sort-of-convinced by the argument and additionally convinced by having several reasonable people ask for `string` now. I don't care strongly either way so I will switch to `string`.

**Comment by dbuenzli** [comment]
> What's wrong with the generic marshaler, exactly?

The result depends on the internal representation. Let's say the algo doesn't change, but the internal machinery does, with a marshal serialization, it's difficult for users to migrate to the new one.

**Comment by gasche** [comment]
> Playing devil's advocate: why add marshaling functions for type Random.State.t and not for other abstract types in the stdlib, e.g. sets or maps?

For sets and maps it's easy to (de)serialize them by going through `key list` or `(key * 'a) list` as a serializable "view" of the opaque type. For the random state, no such view is provided. We could provide a view into a bigarray, would you prefer this? One problem with the bigarray view is that it makes it easier for users to construct PRNG states without using the usual API, and in particular it's less clear whether we should apply the sanitization transformations on the state. (Actually, let me check that the present deserialization function does, I think it should.)

**Comment by gasche** [comment]
So I switched the serialization type to `string`, but I didn't manage to convince myself that `to_string` was a good name (to me this suggests "something you may show to the user", so printability), so I kept the names `to_bytes` and `of_bytes`. What do you think?

**Comment by dbuenzli** [comment]
> What do you think?

That's what I would have done.

**Comment by gasche** [comment]
> We could provide a view into a bigarray, would you prefer this?

Or something else, for example an `int64 list`, which has the advantage of avoiding any mutability/aliasing/separation concern. (But then the view is more costly to build, so for some use-cases it could be less useful.)

File: stdlib/random.mli [hunk 4]
```diff
@@ -148,6 +148,26 @@ module State : sig
...
+  (** Serializes the PRNG state into an immutable sequence of bytes.
+
+      Note that the serialization format may differ across OCaml
+      versions.
+
+      @since 5.1.0
+  *)
+
+  val of_bytes : string -> t
+  (** Deserializes a string obtained by [to_bytes]
```
Comment (dbuenzli) [inline stdlib/random.mli]: `[to_bytes]` -> `{!to_bytes}` but I'm nitpicking.

**Comment by darioteixeira** [comment]
Won't it prove confusing to have functions named `to_bytes` and `of_bytes` when they actually (de)serialise to/from strings? Moreover, since the purpose of these functions is to (de)serialise the state of the RNG, why not just call them `serialize` and `deserialize`?

**Comment by dbuenzli** [comment]
> Moreover, since the purpose of these functions is to (de)serialise the state of the RNG, why not just call them `serialize` and `deserialize`?

I prefer if the functions indicates to/from what they serialize to (e.g. `{to,of}_{json,sexp}` ), rather than a generic term that doesn't tell me much. Even more so that `(de)serialize` is not an established convention in the `Stdlib`. With that respect prefer the current name than what you suggest.

> Won't it prove confusing to have functions named `to_bytes` and `of_bytes` when they actually (de)serialise to/from strings?

I'm not sure the degree of confusion this could lead to when you read the code. In the worst case you might think these are `Bytes.t` but you never see them being mutated by the code, hardly a problem for understanding. 

Personally I got used to use `{to,of}_bytes` with strings. For example to convert abstract hash values to sequences of immutable bytes. As @gasche mentioned above using `to_string` in these cases is confusing due to widespread convention across programming languages this name implies a printable serialisation.

**Comment by dbuenzli** [comment]
Btw. while researching for [that comment](https://github.com/ocaml/ocaml/pull/11329#issuecomment-1159406769) I saw that James Woodyatt uses the term [`octets`](https://ocaml.org/p/orsetto/1.0.3/doc/Cbor_encode/index.html#val-octets). That's a naming convention I could actually be convinced to switch to (that is use `{of,to}_{octets}` when you use strings rather than `{of,to}_bytes`).

**Comment by darioteixeira** [comment]
Yes, I agree that `{of_to}_octets` is a nice name, and a lot less likely to cause confusion!

**Comment by c-cube** [comment]
I've seen `to_bin` / `of_bin` (in the "sha" library I think) which is
also not bad. Or `to_binary` perhaps, to indicate it's not human
readable.

**Comment by xavierleroy** [comment]
>  We could provide a view into a bigarray, would you prefer this? Or something else, for example an int64 list, which has the advantage of avoiding any mutability/aliasing/separation concern.

If you're thinking of interoperability with independent implementations of LXM, perhaps even in other languages, the most useful view is a quadruple (or a record) of int64 (the a, s, x0, x1 components of the state).  Serialization into bytes is the user's responsibility.

**Comment by gasche** [comment]
No, I was not thinking of that. Exposing a quadruple of int64 ties us to the current PRNG algorithm (or others with exactly the same state size), whereas exposing an array/list/bigarray of int64 can be adapted to other formats. The bytes/string serialization I currently propose is actually not such a bad representation either, in particular as it lets us sneak format/version metadata along with the data, so users that want to reason on the specific algorithm used (for interoperability purposes and what not) can do it with reasonable safety.

Let me summarize my current thinking:
- The need expressed by users, which I find reasonable, is to be able to serialize the PRNG state and deserialize again, with enough sanity-checking that it is possible to accept a serialized state from untrusted source. For this going to `bytes` or `string` suffices. This API can remain stable against PRNG algorithm changes, and in fact (cherry on top) we will fail gracefully if passed a serialized state coming from a different algorithm.
- Exposing a more precise view into the data is strictly more expressive, in that it can allow more features to be implemented on top of it without performance overhead, and it keeps the API simple. (But: I'm not sure what other features we want. Comparison/hashing are already supported well.) This said, there is a tension between "low overhead" and mutation-safety: if we don't want to let users mutate the actual PRNG state, we need a copy anyway. (Is it important to prevent this?)

**Comment by xavierleroy** [comment]
> Exposing a quadruple of int64 ties us to the current PRNG algorithm

I'm lost.  I thought the goal was RPC (the multiplayer game example) or more generally interoperability.  You cannot interoperate between different PRNG algorithms.

> exposing an array/list/bigarray of int64 can be adapted to other formats

There's no obvious reason why other PRNGs algorithms would be based on 64-bit integers.

Again, I don't understand what we're trying to do here, and I'm afraid we'll end up with (yet another) poorly designed API in the stdlib.

**Comment by gasche** [comment]
Let me rephrase: we want an API that solves the user need today *and* remains stable if we change the PRNG of the stdlib at some later point.

**Comment by dbuenzli** [comment]
I'm not sure what's so complicated or badly designed about this PR. 

It answers the need to be able to easily continue a sequence started by `Random` at an arbitrary point of its sequence without having to either 1) replay the sequence starting from the seed or 2) resort to tools that may segfault your program or are sensitive to the internal representations of `Random`.

> Remains stable if we change the PRNG of the stdlib at some later point.

You can provide a stable *type* signature, but if you change the PRNG its "data" semantics and effects will change anyway. 

At that point you will have to use @Octachron's future nice compatibility library with your old serialized values. So it's not really a goal in my opinion. 

(And thus I wouldn't even bother with the version number you added to the serialization: as long as the PRNG algorithm is going to be the same it should always be easy to convert the serialization you provide in this PR to whatever the internal representation needs.)

**Comment by gasche** [comment]
On @dbuenzli's advice I have been reviewing my past PRs/issues to wonder if I should close them.

I'm undecided on this one: clearly there is no consensus on the PR, but have the impression that it may due to me not explaining the change clearly enough, rather than a fundamental difference of opinion with @xavierleroy. So I propose to give another shot at explaining the design, and close in a month or so if @xavierleroy is unavailable or remains unconvinced.

## Design

This PR offers a way to serialize/deserialize a Random state safely -- it will fail gracefully on malformed or malicious serialized inputs. As far a I know, the Random API does not currently provide this capability. This could be useful for games, for example -- save a random state in a state file, or exchange it over the network, etc.

The API is abstract enough that we can keep providing it in the future, at the same types, even if the Random implementation changes.

On the other hand, we are explicitly not promising compatibility of serialized values across OCaml versions -- it would be impossible to provide value compatibility across changes to the PRNG algorithm.

In other words, user code using this new API will keep working as intended in future OCaml versions, but only as long as data exchanges remain among compatible OCaml versions -- otherwise a clean error will be reported.

Does that make sense?

**Comment by xavierleroy** [comment]
There's reasonable justification and support for this feature, so I'll stop complaining about it :-)

**Comment by gasche** [comment]
Thanks! Now we would need an approving reviewer if we want to merge -- but of course this is not high-priority.

There remained a naming discussion; there is a consensus that the return type should be `string`, but the name of the functions could be debated. Currently it is `{of,to}_bytes`, which is a reasonable choice for "strings that are data and not meant to be human-readable" but slightly confusing in that the return type is not `bytes`. James Woodyatt uses `{of,to}_octets` for this use-case, and @dbuenzli and @darioteixeira seemed to be in favor. Should I switch to `{of,to}_octets` then?

(I'm familiar with "octet" as a French word to means "byte" or, more precisely, "8-bit integer" (there is a debate on whether bytes are always 8 bits). But I haven't seen it used previously in an English context. No strong opinion on my side, I'm happy to switch to a cute French word if there is consensus.)

**Comment by dbuenzli** [comment]
> Currently it is `{of,to}_bytes`, which is a reasonable choice for "strings that are data and not meant to be human-readable" but slightly confusing in that the return type is not `bytes`

I do have public apis that use `{of,to}_bytes` to return strings (e.g. [here](https://erratique.ch/software/uuidm/doc/Uuidm/index.html#fmt_binary) in `Uuidm`) but I think I will switch to @jhwoodyatt's convention in the future where applicable. It makes thing less confusing to read w.r.t. to the `{of,to}_type` convention (at least until someone defines an `octet` type).

I checked in the ocamldoc index, it seems at the moment all `{of,to}_bytes` functions in the stdlib do take/return `bytes`. So that would set a precedent. `{to,of}_octets` would be a nice convention to establish in the stdlib for the eco-system to copy. (Other possible names would be `Random.State.{encode,decode}` but may be slightly ambiguous to read in this context).

> But I haven't seen it used previously in an English context.

In RFCs I think the term `octet` is actually favoured over `byte` to avoid confusion with non 8-bit bytes. See for example in the UUID RFC [here](https://www.rfc-editor.org/rfc/rfc4122#section-4.1), the [HTTP RFC](https://www.rfc-editor.org/rfc/rfc2616), the [ABNF RFC](https://www.rfc-editor.org/rfc/rfc5234) which defines the non-terminal `OCTET`, etc. That precedent should be enough to counter any accusations of french speaking colonialism.

**Comment by gasche** [comment]
I moved to using `octets`. The `to_octets` documentation currently reads as:

```ocaml
  val to_octets : t -> string
  (** Serializes the PRNG state into an immutable octet sequence.

      The [string] type is intended here for serialization only, the
      encoding is not human-readable and may not be printable.

      Note that the serialization format may differ across OCaml
      versions.

      @since 5.1
  *)
```

**Comment by xavierleroy** [comment]
Apologies for being "that guy" again, but this would be the first occurrence of the word "octet" in the standard library.  Do we really need a new word that sounds like a Gallicism to most of our users to capture the finesse and the je-ne-sais-quoi of your new functions vis-à-vis immutability?

**Comment by c-cube** [comment]
Could it be `{to,of}_bin` or `_binary`? Octets does sound weird and
unusual.

**Comment by gasche** [comment]
I'm fine with whatever that people can agree on.

I think that "octet" is only likely to sound like French to people who already know French. For others I would hazard the  guess that it sounds like "quartet", specialist word but not particularly French.

One mild downside of octet here is that we would be introducing a nuance between two terms "bytes" and "octets" (for mutable and immutable binary payloads), that is difference from the nuance currently established in the computer science community (bytes as encoded characters, octets as groups of binary digits).

File: stdlib/random.ml [hunk 5]
```diff
@@ -40,6 +40,35 @@ module State = struct
     let s = create () in
     set s i1 i2 i3 i4; s
 
+  let serialization_prefix =
+    (* '1' is a version number; more than 200 versions may fit a char. *)
```
Comment (dbuenzli) [inline stdlib/random.ml]: I didn't get that comment. Does that mean that the version is a byte and that the smallest version is 0x49 ?

File: stdlib/random.mli [hunk 6]
```diff
@@ -147,6 +147,29 @@ module State : sig
       correlation.  Both PRNGs can be split later, arbitrarily many times.
       @since 5.0 *)
 
+  val to_octets : t -> string
+  (** Serializes the PRNG state into an immutable octet sequence.
```
Comment (dbuenzli) [inline stdlib/random.mli]: I always prefer when docstrings have the form `[f v] is ...`, but I suspect I get on people's nerve with that obsession. 

Also I would say that you seralize `to` rather than `into` (which I would rather use for an in place modification) but IANANES.

File: stdlib/random.mli [hunk 7]
```diff
@@ -147,6 +147,29 @@ module State : sig
...
+
+      Note that the serialization format may differ across OCaml
+      versions.
+
+      @since 5.1
+  *)
+
+  val of_octets : string -> t
+  (** Deserializes an octet sequence obtained by {!to_octets}
+      into a fresh PRNG state.
```
Comment (dbuenzli) [inline stdlib/random.mli]: Same comments as above.

**Review by dbuenzli — APPROVED** [review]
The PR looks good to me. I added a few nits but feel free to keep things as is.

Regarding the naming story, I think that people that are unhappy with it should at least suggest and justify what they would like to see rather than just complain. FWIW I would also be fine @c-cube's `{to,of}_binary` even though it's less precise :-)

File: stdlib/random.ml (hunk 5 above)
Comment (gasche) [reply stdlib/random.ml]: I don't have a real plan, but we have the option to change this character to indicate incompatible formats of the LXM algorithm, and that there is enough room for many such versions if we want to. (I think we should wait to see the Random algorithm evolve to decide what kind of prefixing/versioning strategy we actually want -- possibly none. In the meantime I like having a self-descriptive format and keeping some wiggle room in it.)

**Comment by xavierleroy** [comment]
> Regarding the naming story, I think that people that are unhappy with it should at least suggest and justify what they would like to see rather than just complain.

My ears are burning.  Personally, I like function names consisting of nouns or verbs that suggest what the function is doing.  `of_<type>` and `to_<type>` aren't terribly evocative, except perhaps for numerical types.  For example, `List.concat` is better than `List.of_list_of_lists`.

So, I would go for `Random.State.serialize` and `Random.State.deserialize`.  Except that @dbuenzli preemptively stroke these names out.   So I'm going to unsubscribe from this discussion and let you choose whatever awful `of_<type>/to_<type>` names you want.

**Comment by jhwoodyatt** [comment]
>Apologies for being "that guy" again, but this would be the first occurrence of the word "octet" in the standard library. Do we really need a new word that sounds like a Gallicism to most of our users to capture the finesse and the je-ne-sais-quoi of your new functions vis-à-vis immutability?

I don't think "octet" retains any special Gallic flavor anymore. Consider the results of a Google search for instances of the word on the `ietf.org` web site: [link](https://www.google.com/search?q=octet+site%3Aietf.org).

p.s. Fun fact, the IETF adopted "octet" pretty easily, but it has for a long time had a naming problem for 16-bit quantities, which are most commonly found in the colon-separated fields of IPv6 address. At some point, a discussion erupted in the IPv6 working group about the various competing terms that have been used in various drafts and standards track documents. Rough consensus was not to be found. I got very bored one weekend and did some research into the historical origins of "octet" in the hopes that I could find a good candidate that would break the impasse. I discovered to my delight that while the word was coined by French people, the word is basically an import from Italian. (It came from the Italian words for musical groups of specific number of intruments.) Alas, Italian does not actually have a word for a group with exactly sixteen musical instruments. (Because why would you need one?) So, I consulted my friends who are professional English-Italian translators and coined my own word for a 16-bit quantity: **sedicet**. I was very proud of myself, and my translator friends were very happy with how it sounds, but alas, by then the controversy in the IPv6 working group petered out without any productive result, and I was hesitant to reopen it. I still think "sedicet" is an awesome word, and we should all be using it.

**Comment by shindere** [comment]
Regarding the naming, @damiendoligez has suggested `of/to_blob`.

He emmphasized that the meaning of `blob` is already known, at least
in the database area, as mmmeaning `Binary Large OBject`, so it's
maybe a bit more precise (well, let's say less vague than `data` in the
sense that it conveys the binary (thus non rintable) nature of the
object. It's perhaps not a large object, but maybe the name is still
good enough?

**Comment by dbuenzli** [comment]
> Regarding the naming, @damiendoligez has suggested `of/to_blob`.

Note that in browsers [`Blob` objects](https://developer.mozilla.org/en-US/docs/Web/API/Blob) are a thing. So you naturally get `{of,to}_blob` in JavaScript bindings (e.g. [[1]](https://erratique.ch/software/brr/doc/Brr_canvas/Canvas/index.html#val-to_blob),[[2]](https://erratique.ch/software/brr/doc/Brr/File/index.html#val-of_blob), [[3]](https://erratique.ch/software/brr/doc/Brr_io/Fetch/Body/index.html#val-of_blob)).

**Comment by nojb** [comment]
I don't really want to step into this ring, but I don't think `to_bytes` and `of_bytes` is a good fit here. (My first guess would be that `Random.State.of_bytes` means a state seeded with some bytes).

In `Digest` we have `{to,of}_hex` for the serialization functions, what about `{to,of}_binary` or `{to,of}_bin` here? (This has already been suggested above.)

**Comment by gasche** [comment]
Another idea I came up during lunch was `{of,to}_binary_string`. We discussed `{,de}serialize` as well, I like the idea but I am hesitant about the fact that none of the existing conversion function use this approach (and I view it as a larger break from conventions than inventing a new cute name for use after `{of,to}_`); Damien remarked that it is a bit awkward that for many people "marshal" and "serialize" are synonyms, while we would introduce a subtle distinction. He also proposed `{of,to}_data`.

Finally @shindere proposed having a different type `Bstring.t` for "binary strings", which would make things more clear/explicit. @Octachron then counter-attacked that it should be the other way around, have a `Text.t` for "strings intended for display". Nobody was fond of my suggestion to have both and leave `String.t` as a muddy border zone. After toying with this a bit, I think that it could make sense to introduce a *synonym* of `String.t`, `String.bin` or `String.binary` or `bstring`, and gradually change our APIs to use it to indicate the nuance. This does not provide a forward-compatible path to actually making the type distincts (well, we did start this way with `bytes` after all), but I don't know that we ever will and perhaps the sweet&sour fragility, the quaint smell and weird inference behavior of type synonyms is what we need here.

To summarize:
- I like `{of,to}_{bin,data,bstring,binary,binary_string}`. For some reason I prefer `_bin` and `_binary_string` to `_bstring` or `_binary`. (To me "a binary" is the data representing a program; `_bin` could mean more and `_binary_string` is more explicit. I don't mind the extra length when it makes the code clearer to read.)
- This gave me ideas about having a type of that name to make things clearer also in the types, but I would rather leave this question to a different discussion thread.

**Comment by gasche** [comment]
Remark: After `_hex`, it would also be natural to suggest `_bits`, maybe even more natural than `_bin`. We already have `Int64.float_of_bits`, that is used in a non-incompatible way (it suggests the interpretation of another type, here `int64`, just a sequence of bits).

**Comment by dbuenzli** [comment]
I think we should rather not add more types or type aliases, it will likely introduce more confusion in the system. But rather stick and teach the following model:

0. `Char.t` is a representation for bytes.
1. `Bytes.t` is a mutable sequence of bytes.
2. `String.t` is an immutable sequence of bytes.
3. If `String.t` is to be interpreted as text then *unless otherwise noted* it should be UTF-8 encoded text.
4. `M.{to,of}_t` represent conversions from `M.t` to a type or format `t`.

Now the problem of 4. and the `string` type is that people do naturally read into `M.{to,of}_string` the interpretation 3 and so expect the result to be usable in text output. We need a convention for the eco-system to distinguish between 2. and 3. I propose that convention to be:

1. `M.{to,of}_string` for the UTF-8 textual representation.
2. `M.{to,of}_binary_string` for the immutable sequence of bytes representation.

**Comment by gasche** [comment]
I rebased the PR to use `binary_string` instead of `octet`. Thanks!

While I was at it, I asked myself if the error message on invalid format could be more informative than

```ocaml
      failwith "Random.State.of_binary_string"
```

I pushed a new commit that uses instead:

```ocaml
      Printf.ksprintf failwith
        "Random.State.of_binary_string: invalid format (expected version: %s)"
        Sys.ocaml_version;
```

(I haven't figured out yet how to un-break my build that seems to be not expecting a dependency of Random on Printf. `make depend` seems to do nothing. Build systems are a mess.)

Is there a good reason to absolutely not try to be helpful in this way?

**Comment by shindere** [comment]
Quick question: why don't you raise `Invalid_argument` on failure?

**Comment by gasche** [comment]
Invalid_argument signals a programming error, the caller should have checked against this case in the first place. Failure is for other failure conditions, possibly coming from unsupported inputs.

**Comment by dbuenzli** [comment]
+1 for better error messages 

> (I haven't figured out yet how to un-break my build that seems to be not expecting a dependency of Random on Printf. `make depend` seems to do nothing. Build systems are a mess.)

I think this is https://github.com/ocaml/ocaml/issues/11588

File: stdlib/random.ml [hunk 8]
```diff
@@ -40,6 +40,37 @@ module State = struct
...
+
+  let of_binary_string buf =
+    let prefix = serialization_prefix in
+    let preflen = serialization_prefix_len in
+    if String.length buf <> preflen + (4 * 8)
+       || not (String.starts_with ~prefix buf)
+    then
+      Printf.ksprintf failwith
+        "Random.State.of_binary_string: invalid format (expected version: %s)"
+        Sys.ocaml_version;
```
Comment (dbuenzli) [inline stdlib/random.ml]: However since multiple ocaml versions are going to share the same formt this is slightly misleading. Why not use:
```
  "Random.State.of_binary_string: invalid format (expected %s)" serialization_prefix;
```

Comment (gasche) [reply stdlib/random.ml]: Thanks for the feedback.

My intuition is that users have no clue about the serialization format they are producing and/or should be expecting, but that they may have a sense of which OCaml version they expect to be running, and understand what is going on if they see a different version from that.

Maybe we could say "expected a format compatible with {version}"?

This being said:
- I'm making the assumption that the issue is related to the version of the running program. (The code could be incorrect in other ways, or consume serialized data from another version, etc.)
- The hint also impresses this assumption into the users and may lead them in false directions
- I don't have the energy to find the best approach right now, so I think that I will just fall back to the previous implementation. (Also: maybe I will be able to build the branch locally again.)

Comment (dbuenzli) [reply stdlib/random.ml]: > Maybe we could say "expected a format compatible with {version}"?

Maybe that's good enough. That way you can also simply:
```
failwith ("Random.State.of_binary_string: expected a format compatible with OCaml " ^ Sys.ocaml_version)
```
and get rid of your `.depend` woes.

Comment (gasche) [reply stdlib/random.ml]: Barely, thanks to the fact that Int32 and Nativeint already depend on Sys. Phew! Thanks, I will take your suggestion.

**Comment by shindere** [comment]
Gabriel Scherer (2023/01/11 07:20 -0800):
> (I haven't figured out yet how to un-break my build that seems to be
> not expecting a dependency of Random on Printf. `make depend` seems to
> do nothing. Build systems are a mess.)

See #11885 for a proposed fix.

**Comment by gasche** [comment]
I have gone over the existing review comments, and clarified the implementation and documentation comments to ensure that they now answer all the questions asked by the reviewers.

I think this PR is ready for a new round of review and hopefully an approval.

**Review by dbuenzli — APPROVED** [review]

**Review by nojb — APPROVED** [review]
To be honest, I am not really convinced by `{to,of}_binary_string`, but I don't think it is worth our time to debate this endlessly. Approved (mostly in terms of @dbuenzli's review, I have only glanced at the code.)

**Comment by nojb** [comment]
We will need a second official approval to merge this PR. @dra27?

File: stdlib/random.ml [hunk 9]
```diff
@@ -40,6 +40,48 @@ module State = struct
...
+    Bytes.blit_string prefix 0 buf 0 preflen;
+    for i = 0 to 3 do
+      Bytes.set_int64_le buf (preflen + i * 8) (Array1.get s i)
+    done;
+    Bytes.unsafe_to_string buf
+
+  let of_binary_string buf =
+    let prefix = serialization_prefix in
+    let preflen = serialization_prefix_len in
+    if String.length buf <> preflen + (4 * 8)
```
Comment (damiendoligez) [inline stdlib/random.ml]:
```suggestion
    if String.length buf <> preflen + 4 * 8
```

File: stdlib/random.mli [hunk 10]
```diff
@@ -147,6 +147,41 @@ module State : sig
...
+      versions.
+
+      @since 5.1
+  *)
+
+  val of_binary_string : string -> t
+  (** Deserializes a byte sequence obtained by {!to_binary_string}.
+      The resulting PRNG state should produce exactly the same random
+      numbers as the state that was passed in input to
+      {!to_binary_string}.
```
Comment (damiendoligez) [inline stdlib/random.mli]:
```suggestion
  (** Deserializes a byte sequence obtained by calling {!to_binary_string}.
      The resulting PRNG state will produce the same random
      numbers as the state that was passed as input to
      {!to_binary_string}.
```

File: stdlib/random.mli [hunk 11]
```diff
@@ -147,6 +147,41 @@ module State : sig
...
+      versions.
+
+      Unlike the functions provided by the {!Marshal} module, this
+      function either produces a valid state or fails cleanly with
+      a [Failure] exception. It can be safely used on user-provided,
+      untrusted inputs.
+
+      It could be useful, for example, to store the current random
+      state in the save file of a video game, or to have players
+      communicate their random state over the network.
```
Comment (damiendoligez) [inline stdlib/random.mli]: I don't think we need this paragraph.

**Review by damiendoligez — APPROVED** [review]
I only have a few nits. Feel free to ignore them.

`binary_string` seems to be a good compromise: it's acceptable to everyone but nobody likes it...

File: stdlib/random.mli (hunk 11 above)
Comment (gasche) [reply stdlib/random.mli]: @damiendoligez @dbuenzli, I would be curious to know just a little more about your impression here. I thought on the contrary that it was sensible to try to distill in the documentation the sort of use-cases that convinced us (me at least?) to add the feature in the first place. (People often mention that they want to see examples in the documentation. This is not a code example but a usage example, and I thought it could be useful.) I would be curious to hear your opinion on this point to inform my future documentation-writing practice.

Comment (dbuenzli) [reply stdlib/random.mli]: Maybe it's style but I don't like chatty reference docs. I find these kind of suggestions a bit noisy and in the way. This is the kind of stuff you rather have in tutorials/howto docs. 

I rarely peruse reference documentation in search of use-cases, I rather peruse it knowing what I'm looking for. As such I want docstrings to communicate me efficiently how to use the structure item, possible pitfalls, pointers to nudge me to alternatives (where task based suggestions are ok) or towards more extensive docs. Nothing more distracting.

Keeping doc strings to the point also enables more efficient scanning of a module's offering (this is the reason why I often defer via a link a long module preamble or extensive explanations to a section at the end of the module)

**Comment by gasche** [comment]
I integrated the feedback of @damiendoligez. This should be good to merge once the CI agrees.

**Comment by gasche** [comment]
thanks @nojb :-)
